modules/freestarBidAdapter.js

```diff
@@ -0,0 +1,209 @@
+import { _each, isFn, isPlainObject, isNumber, isStr, deepAccess, parseUrl, _map, getUniqueIdentifierStr, createTrackPixelHtml } from '../src/utils.js';
+import { registerBidder } from '../src/adapters/bidderFactory.js';
+import { config } from '../src/config.js';
+import { BANNER } from '../src/mediaTypes.js';
+import { createEidsArray } from './userId/eids.js';
+
+const BIDDER_CODE = 'freestar';
+const SERVER_URL = 'https://c.deployads.com';
+
+function transformSyncs(responses, type, syncs) {
+  _each(responses, res => {
+    if (res.body && res.body.ext && res.body.ext.sync_dsps && res.body.ext.sync_dsps.length) {
+      _each(res.body.ext.sync_dsps, sync => {
+        if (sync[0] === type && sync[1]) {
+          syncs.push({type, url: sync[1]});
+        }
+      });
+    }
+  });
+}
+
+function getBidFloor(bid) {
+  if (!isFn(bid.getFloor)) {
+    return bid.params.floor ? bid.params.floor : null;
+  }
+
+  // MediaType and Size will automatically get set for us if the bid only has
+  // one media type or one size.
+  let floor = bid.getFloor({
+    currency: 'USD',
+    mediaType: '*',
+    size: '*'
+  });
+  if (isPlainObject(floor) && !isNaN(floor.floor) && floor.currency === 'USD') {
+    return floor.floor;
+  }
+  return null;
+}
+
+export const spec = {
+  code: BIDDER_CODE,
+  supportedMediaTypes: [BANNER],
+
+  isBidRequestValid: function(bid) {
+    const fsConfig = config.getConfig('freestar');
+    const haveSiteId = (fsConfig && !!fsConfig.siteId) || bid.params.siteId;
+    const floor = getBidFloor(bid);
+    const validFloor = !floor || isNumber(floor);
+    const validKeywords = !bid.params.keywords ||
+      (isPlainObject(bid.params.keywords) &&
+        Object.keys(bid.params.keywords).every(key =>
+          isStr(key) && isStr(bid.params.keywords[key])
+        ))
+    const isBanner = !bid.mediaTypes || bid.mediaTypes[BANNER];
+    const bannerSizes = isBanner ? deepAccess(bid, `mediaType.${BANNER}.sizes`) || bid.sizes : null;
+    return !!(bid.params.tagId && haveSiteId && validFloor && validKeywords && (!isBanner ||
```

Comment:
are you really rejecting biid requests if the keywords are invalid? why not just drop the keywords

Reply:
Yes, we are validating just as we did in the SortableBidAdapter.

Reply:
this validation is quite aggressive, doesnt seem best for your company or your publisher partners

Reply:
also why not fallback to no floor when the floor is invalid?